components/psst/browser/core/psst_rule.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/browser/core/psst_rule.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/files/file_path.h"
+#include "base/json/json_reader.h"
+#include "base/types/expected.h"
+#include "extensions/common/url_pattern.h"
+#include "net/base/url_util.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace {
+
+// Format:
+// [
+//   {
+//     "include": [
+//       "*://twitter.com"
+//     ],
+//     "exclude": [
+//     ],
+//     "version": 1,
+//     "test_script": "twitter/test.js",
+//     "policy_script": "twitter/policy.js"
+//   }, ...
+// ]
+// Note that "test_script" and "policy_script" give paths
+// relative to the component under scripts/
+
+// psst.json keys
+const char kInclude[] = "include";
+const char kExclude[] = "exclude";
+const char kTestScript[] = "test_script";
+const char kPolicyScript[] = "policy_script";
+const char kVersion[] = "version";
+
+}  // namespace
+
+namespace psst {
+
+PsstRule::PsstRule() = default;
+
+PsstRule::~PsstRule() = default;
+
+// static
+bool PsstRule::GetURLPatternSetFromValue(const base::Value* value,
+                                         extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(
+      value->GetList(), URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS,
```

Comment:
Let's drop `SCHEME_HTTP` since I can't think of a site we would want to do that's still using legacy HTTP.